xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-default/src/main/java/org/xwiki/notifications/notifiers/internal/UserEventDispatcher.java

```diff
@@ -116,165 +114,178 @@ public class UserEventDispatcher implements Runnable, Disposable, Initializable
-        // will fallback on the large queue. If both queue are full we just ignore the event and it will be picked up by
-        // the next missing pre filtered event run.
-        this.priorityQueue = new LinkedBlockingQueue<>(1000);
-        this.secondaryQueue = new LinkedBlockingQueue<>(100000);
-
-        // Start a background thread to filter and dispatch users events
-        Thread thread = new Thread(this);
-        thread.setName("User event dispatcher thread");
-        thread.setPriority(Thread.NORM_PRIORITY - 1);
-        thread.setDaemon(true);
-        thread.start();
-    }
+    @Inject
+    private RemoteObservationManagerConfiguration remoteObservation;
 
-    @Override
-    public void dispose() throws ComponentLifecycleException
-    {
-        // Mark the component as disposed
-        this.disposed = true;
+    @Inject
+    private ExecutionContextManager contextManager;
 
-        // Make sure to wake the thread up to stop it
-        wakeup();
-    }
+    @Inject
+    private Execution execution;
 
-    private void wakeup()
+    @Inject
+    private Logger logger;
+
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+
+    private volatile boolean running;
+
+    /**
+     * Start the scheduler.
+     */
+    public void initialize()
     {
-        if (this.priorityQueue != null && this.priorityQueue.isEmpty()) {
-            // We don't need to wait since we just want to make sure that the queue is not empty
-            this.priorityQueue.offer(WAKEUP_EVENT);
-        }
+        // Schedule a job to regularly check if any event prefiltering was missed or a previous run failed
+        this.scheduler.scheduleWithFixedDelay(this::run, 0, 1, TimeUnit.HOURS);
     }
 
     /**
-     * @param event the event to dispatch
-     * @return true if the event was added to the queue, false if it was ignored (in which case it will generally
-     *         retried later by {@link PrefilterMissingEventsJob})
-     * @throws InterruptedException if interrupted while waiting
+     * Indicate an event just been created.
      */
-    public boolean addEvent(Event event) throws InterruptedException
+    public void onEvent(Event event)
     {
-        if (!this.disposed) {
-            // Put in the priority queue only live events (event which just been produced)
-            if (event.getDate() == null
-                || (System.currentTimeMillis() - event.getDate().getTime()) >= LIVE_EVENT_INTERVAL
-                || !this.priorityQueue.offer(event)) {
-                return addEvent(event.getId(), false);
+        try {
+            if (getSupportedEventTypes().contains(event.getType()) && !this.running) {
+                // Make sure to wakeup the dispatcher
+                this.scheduler.execute(this::run);
             }
-
-            return true;
+        } catch (EventStreamException e) {
+            this.logger.error("Failed to get supported event types", e);
```

Comment:
Does the root error contains the info of the event type in such case? Else it would worth to provide it in the error message probably.

Reply:
Not sure what you mean, when you end up here what failed is the loading of the supported event types in general, and it's not related to a specific one.

Reply:
Right, thought it was related to the event so all good.